pilot/pkg/bootstrap/server.go

```diff
@@ -944,3 +889,114 @@ func (s *Server) getIstiodCertificate(info *tls.ClientHelloInfo) (*tls.Certifica
 	return s.istiodCert, nil
 }
+
+// parseIstiodAddress parses the Istiod address and validates it.
+func (s *Server) parseIstiodAddress() (string, string, error) {
+	istiodAddr := features.IstiodService.Get()
+	if istiodAddr == "" {
+		// Feature disabled
+		return "", "", nil
+	}
+
+	// validate
+	host, port, err := net.SplitHostPort(istiodAddr)
+	if err != nil {
+		return "", "", fmt.Errorf("invalid ISTIOD_ADDR(%s): %v", istiodAddr, err)
+	}
+	if _, err := strconv.Atoi(port); err != nil {
+		return "", "", fmt.Errorf("invalid port(%s) in ISTIOD_ADDR(%s): %v", port, istiodAddr, err)
+	}
+
+	return host, port, nil
+}
+
+// initControllers initializes the controllers.
+func (s *Server) initControllers(args *PilotArgs) error {
+	// Certificate controller is created before MCP controller in case MCP server pod
+	// waits to mount a certificate to be provisioned by the certificate controller.
+	if err := s.initCertController(args); err != nil {
+		return fmt.Errorf("error initializing certificate controller: %v", err)
+	}
+	if err := s.initConfigController(args); err != nil {
+		return fmt.Errorf("error initializing config controller: %v", err)
+	}
+	if err := s.initServiceControllers(args); err != nil {
+		return fmt.Errorf("error initializing service controllers: %v", err)
+	}
+	return nil
+}
+
+// initNamespaceController initializes namespace controller to sync config map.
+func (s *Server) initNamespaceController(args *PilotArgs) {
+	if s.ca != nil && s.kubeClient != nil {
+		s.addTerminatingStartFunc(func(stop <-chan struct{}) error {
+			leaderelection.
+				NewLeaderElection(args.Namespace, args.PodName, leaderelection.NamespaceController, s.kubeClient).
+				AddRunFunction(func(stop <-chan struct{}) {
+					log.Infof("Starting namespace controller")
+					nc := kubecontroller.NewNamespaceController(s.fetchCARoot, args.Config.ControllerOptions, s.kubeClient)
+					nc.Run(stop)
+				}).
+				Run(stop)
+			return nil
+		})
+	}
+}
+
+// initGenerators initializes generators to be used by XdsServer.
+func (s *Server) initGenerators() {
+	s.EnvoyXdsServer.Generators["grpc"] = &grpcgen.GrpcConfigGenerator{}
+	s.EnvoyXdsServer.Generators["grpc/"+envoyv2.EndpointType] = &envoyv2.EdsGenerator{Server: s.EnvoyXdsServer}
+}
+
+// initJwtPolicy initializes JwtPolicy.
+func (s *Server) initJwtPolicy() {
+	if features.JwtPolicy.Get() != jwt.JWTPolicyThirdPartyJWT {
+		log.Infoa("JWT policy is ", features.JwtPolicy.Get())
+	}
+
+	switch features.JwtPolicy.Get() {
+	case jwt.JWTPolicyThirdPartyJWT:
+		s.jwtPath = ThirdPartyJWTPath
+	case jwt.JWTPolicyFirstPartyJWT:
+		s.jwtPath = securityModel.K8sSAJwtFileName
+	default:
+		log.Infof("unknown JWT policy %v, default to certificates ", features.JwtPolicy.Get())
+	}
+}
+
+// maybeCreateCA creates and initializes CA Key if needed.
+func (s *Server) maybeCreateCA(args *PilotArgs, caOpts *CAOptions) error {
+	// CA signing certificate must be created only if custom TLS certs are not provided.
```

Comment:
This doesn't sound right. If I add an ACME certificate for Istiod - which is very reasonable for MC and mesh expansion use cases - it should not disable Istiod CA functionality. 

I guess not related to your refactoring - but someone else, but it is a problem.

Reply:
Are you saying  if `caBundlePath` is specified we should allow CA creation - There is a TODO to unify caBundlePath and CACert in TLS options. I will clean that up as part of that work.

Reply:
I think this means that we may want to pass in our own certs to use for the webhook/XDS, but we still need the CA functionality running so that we can sign certs for workloads. I think in your case you don't use any CA functionality but there is potential use case to use different certs for the control plane but still use Istiod as a CA?

Reply:
>  but there is potential use case to use different certs for the control plane but still use Istiod as a CA?

Ok.  That is fine. Should I just base it on `ENABLE_CA_SERVER` ? That makes sense to me actually

Reply:
Fixed in https://github.com/istio/istio/pull/23698/commits/d9de48491310f62017982e5b0a292b8dc6743958